### PR TITLE
Correctly handling __proto__ export as module object key

### LINKED ISSUE #3042: dispose __proto__ export properly
- Rollup Version: 1.19.4
- Operating System (or Browser):
- Node Version:

### How Do We Reproduce?

rollup options:

```toml
output.format = 'cjs'
```

main.js:

```js
import * as y from './mod.js';
export var x = 0;
export var __proto__ = y;
export var z = 2;
```

mod.js:

```js
export var a = 0;
export var __proto__ = 1;
export var b = 2;
```

### Expected Behavior

output[0].code:

```js
var a = 0;
var __proto__ = 1;
var b = 2;

var y = {
    a: a,
    __proto__: null,// case 2
    b: b
};
y.__proto__ = __proto__;

exports.x = 0;
exports.__proto__ = null;// case 1
exports.__proto__ = y;
exports.z = 2;
```

### Actual Behavior

output[0].code:

```js
var a = 0;
var __proto__ = 1;
var b = 2;

var y = {
    a: a,
    __proto__: __proto__,
    b: b
};

exports.x = 0;
exports.__proto__ = y;
exports.z = 2;
```

**case 1**: Set property `__proto__` to `null` before really set it would be safer.

**case 2**: Declare property `__proto__` as `null`, and set it later.

### PR DESCRIPTION
**Pull request by LongTengDao**
This PR contains:

- [x] bugfix
- [ ] feature
- [ ] refactor
- [ ] documentation
- [ ] other

Are tests included?

- [x] yes (_bugfixes and features will not be merged without tests_)
- [ ] no

Breaking Changes?

- [ ] yes (_breaking changes will not be merged unless absolutely necessary_)
- [x] no

List any relevant issue numbers:

-  resolves #3042

### Description

https://rollupjs.org/repl/?version=4.9.1&shareable=JTdCJTIyZXhhbXBsZSUyMiUzQW51bGwlMkMlMjJtb2R1bGVzJTIyJTNBJTVCJTdCJTIyY29kZSUyMiUzQSUyMmltcG9ydCUyMColMjBhcyUyMHglMjBmcm9tJTIwJy4lMkZ4JyUzQiU1Q25jb25zb2xlLmxvZyh4KSUzQiUyMiUyQyUyMmlzRW50cnklMjIlM0F0cnVlJTJDJTIybmFtZSUyMiUzQSUyMm1haW4uanMlMjIlN0QlMkMlN0IlMjJjb2RlJTIyJTNBJTIyZXhwb3J0JTIwY29uc3QlMjBfX3Byb3RvX18lMjAlM0QlMjAwJTNCJTIyJTJDJTIyaXNFbnRyeSUyMiUzQWZhbHNlJTJDJTIybmFtZSUyMiUzQSUyMnguanMlMjIlN0QlNUQlMkMlMjJvcHRpb25zJTIyJTNBJTdCJTIyb3V0cHV0JTIyJTNBJTdCJTIyZm9ybWF0JTIyJTNBJTIyZXMlMjIlN0QlMkMlMjJ0cmVlc2hha2UlMjIlM0F0cnVlJTdEJTdE

The PR generate `["__proto__"]:` instead of `__proto__:`. It can also be written by `get __proto__ ()` instead, if that would be better. All formats have no `__proto__` related problem now.

### CODE REVIEW

**Comment by LongTengDao** [comment]
no idea about failed test

[Build x86_64-unknown-linux-gnu](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#logs)

```
Build in docker

Run addnab/docker-run-action@v3
  with:
    image: ghcr.io/napi-rs/napi-rs/nodejs-rust:lts-debian
    options: --user 0:0 -v /home/runner/work/rollup/rollup/.cargo-cache/git/db:/usr/local/cargo/git/db -v /home/runner/work/rollup/rollup/.cargo/registry/cache:/usr/local/cargo/registry/cache -v /home/runner/work/rollup/rollup/.cargo/registry/index:/usr/local/cargo/registry/index -v /home/runner/work/rollup/rollup:/build -w /build
    run: set -e && rustup target add x86_64-unknown-linux-gnu && npm run build:napi -- --release --target x86_64-unknown-linux-gnu && strip *.node
    shell: sh
  env:
    DEBUG: napi:*
    APP_NAME: rollup
    MACOSX_DEPLOYMENT_TARGET: 10.13
/usr/bin/docker run --name de841ba3d63[2](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:2)48220e4c8cab89e04cfebbd4af_646e5e --label de841b --workdir /github/workspace --rm -e "DEBUG" -e "APP_NAME" -e "MACOSX_DEPLOYMENT_TARGET" -e "INPUT_IMAGE" -e "INPUT_OPTIONS" -e "INPUT_RUN" -e "INPUT_SHELL" -e "INPUT_REGISTRY" -e "INPUT_USERNAME" -e "INPUT_PASSWORD" -e "INPUT_DOCKER_NETWORK" -e "HOME" -e "GITHUB_JOB" -e "GITHUB_REF" -e "GITHUB_SHA" -e "GITHUB_REPOSITORY" -e "GITHUB_REPOSITORY_OWNER" -e "GITHUB_REPOSITORY_OWNER_ID" -e "GITHUB_RUN_ID" -e "GITHUB_RUN_NUMBER" -e "GITHUB_RETENTION_DAYS" -e "GITHUB_RUN_ATTEMPT" -e "GITHUB_REPOSITORY_ID" -e "GITHUB_ACTOR_ID" -e "GITHUB_ACTOR" -e "GITHUB_TRIGGERING_ACTOR" -e "GITHUB_WORKFLOW" -e "GITHUB_HEAD_REF" -e "GITHUB_BASE_REF" -e "GITHUB_EVENT_NAME" -e "GITHUB_SERVER_URL" -e "GITHUB_API_URL" -e "GITHUB_GRAPHQL_URL" -e "GITHUB_REF_NAME" -e "GITHUB_REF_PROTECTED" -e "GITHUB_REF_TYPE" -e "GITHUB_WORKFLOW_REF" -e "GITHUB_WORKFLOW_SHA" -e "GITHUB_WORKSPACE" -e "GITHUB_ACTION" -e "GITHUB_EVENT_PATH" -e "GITHUB_ACTION_REPOSITORY" -e "GITHUB_ACTION_REF" -e "GITHUB_PATH" -e "GITHUB_ENV" -e "GITHUB_STEP_SUMMARY" -e "GITHUB_STATE" -e "GITHUB_OUTPUT" -e "RUNNER_OS" -e "RUNNER_ARCH" -e "RUNNER_NAME" -e "RUNNER_ENVIRONMENT" -e "RUNNER_TOOL_CACHE" -e "RUNNER_TEMP" -e "RUNNER_WORKSPACE" -e "ACTIONS_RUNTIME_URL" -e "ACTIONS_RUNTIME_TOKEN" -e "ACTIONS_CACHE_URL" -e "ACTIONS_RESULTS_URL" -e GITHUB_ACTIONS=true -e CI=true -v "/var/run/docker.sock":"/var/run/docker.sock" -v "/home/runner/work/_temp/_github_home":"/github/home" -v "/home/runner/work/_temp/_github_workflow":"/github/workflow" -v "/home/runner/work/_temp/_runner_file_commands":"/github/file_commands" -v "/home/runner/work/rollup/rollup":"/github/workspace" de841b:a[3](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:3)d632[4](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:4)8220e4c8cab89e04cfebbd4af
Unable to find image 'ghcr.io/napi-rs/napi-rs/nodejs-rust:lts-debian' locally
lts-debian: Pulling from napi-rs/napi-rs/nodejs-rust
docker: Get "https://ghcr.io/v2/napi-rs/napi-rs/nodejs-rust/manifests/sha2[5](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:5)[6](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:6):d8bf835a3a8a8[7](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:7)3c05[8](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:8)d0528[9](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:9)4d83706a90684ea9ee2f62[10](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:10)f75c6bfcd4800d3": dial tcp 140.82.[11](https://github.com/rollup/rollup/actions/runs/7334340036/job/19970891775?pr=5313#step:11:12)2.33:443: i/o timeout.
See 'docker run --help'.
```

**Comment by LongTengDao** [comment]
@lukastaegert ready to merge. sorry for wasting too much time to deal with existing tests with hash...

File: src/utils/systemJsRendering.ts [hunk 1]
```diff
@@ -16,7 +16,8 @@ export function getSystemExportStatement(
 			exportNamesByVariable.get(variable)![0]
 		)},${_}${variable.getName(getPropertyAccess)}${modifier})`;
 	} else {
-		const fields: [key: string, value: string][] = [];
+		const fields: [key: string | null, value: string][] = [];
+		fields.push([null, `__proto__:${_}null`]);
```
Comment (lukastaegert) [inline src/utils/systemJsRendering.ts:20]: I do not unterstand why we need this. This generates a lot of noise and if we set a potential __proto__ variable via `["__proto__"]`, this should not have any effect.

File: src/ast/nodes/ExportDefaultDeclaration.ts [hunk 2]
```diff
@@ -146,7 +146,9 @@ export default class ExportDefaultDeclaration extends NodeBase {
 			code.overwrite(
 				this.start,
 				declarationStart,
-				`${cnst} ${this.variable.getName(getPropertyAccess)} = exports('${systemExportNames[0]}', `
+				`${cnst} ${this.variable.getName(getPropertyAccess)} = exports(${JSON.stringify(
+					systemExportNames[0]
+				)}, `
```
Comment (lukastaegert) [inline src/ast/nodes/ExportDefaultDeclaration.ts:151]: I included this change in master now as it is not central to this PR and makes the diff much smaller (also, I do not think it is strictly necessary except to have some symmetry in what quotes are used for exports=.

**Comment by lukastaegert** [comment]
BTW you can use `npm run test:update-snapshots` to update most snapshots automatically.

**Comment by LongTengDao** [comment]
@lukastaegert Under your patient guidance, the heavy maintenance work suddenly became enjoyable!

**Review by lukastaegert — APPROVED** [review]

**Pull request merged**